File: sample/client/ts/engine/get_rank_cli.ts
import * as engine from "./engine";
import { encode, decode } from "./@msgpack/msgpack";
import * as common from "./common_cli";
// this enum code is codegen by geese codegen for ts

// this struct code is codegen by geese codegen for ts
// this caller code is codegen by geese codegen for typescript
export class get_rank_get_self_rank_cb {
    public entity:engine.subentity;
    public cb:(rank:common.role_rank_info) => void;
    public err:(err:common.error_code) => void;
    public rsp:engine.callback;
    public constructor(_cb_uuid:number, _entity:engine.subentity|engine.player) {
        this.entity = _entity
        this.rsp = new engine.callback(() => { this.entity.del_callback(_cb_uuid); });
        this.entity.reg_hub_callback(_cb_uuid, this.rsp)

    }

    private on_rsp(bin:Uint8Array) {
        let inArray = decode(bin) as any;
        let _rank = common.protcol_to_role_rank_info(inArray[0]);
        this.cb(_rank);

    }

    private on_err(bin:Uint8Array) {
        let inArray = decode(bin) as any;
        let _err = inArray[0];
        this.err(_err)

    }

    public  callBack(_cb:(rank:common.role_rank_info) => void, _err:(err:common.error_code) => void) {
        this.cb = _cb;
        this.err = _err;
        this.rsp.callback(this.on_rsp, this.on_err);
        return this.rsp;
    }

}

export class get_rank_get_rank_cb {
    public entity:engine.subentity;
    public cb:(rank_list:Array<common.role_rank_info>) => void;
    public err:(err:common.error_code) => void;
    public rsp:engine.callback;
    public constructor(_cb_uuid:number, _entity:engine.subentity|engine.player) {
        this.entity = _entity
        this.rsp = new engine.callback(() => { this.entity.del_callback(_cb_uuid); });
        this.entity.reg_hub_callback(_cb_uuid, this.rsp)

    }

    private on_rsp(bin:Uint8Array) {
        let inArray = decode(bin) as any;
        let _rank_list:Array<common.role_rank_info> = [];
        for (let v_e249ecbd_d64c_526b_901b_6d4ddee4b75a of inArray[0]) {
            _rank_list.push(common.protcol_to_role_rank_info(v_e249ecbd_d64c_526b_901b_6d4ddee4b75a))
        }
        this.cb(_rank_list);

    }

    private on_err(bin:Uint8Array) {
        let inArray = decode(bin) as any;
        let _err = inArray[0];
        this.err(_err)

    }

    public  callBack(_cb:(rank_list:Array<common.role_rank_info>) => void, _err:(err:common.error_code) => void) {
        this.cb = _cb;
        this.err = _err;
        this.rsp.callback(this.on_rsp, this.on_err);
        return this.rsp;
    }

}

export class get_rank_caller {
    public entity:engine.subentity|engine.player;
    public constructor(entity:engine.subentity|engine.player) {
        this.entity = entity;
    }

    public  get_self_rank(entity_id:string) {
        let _argv_e22ae90d_2428_3197_a8fb_549203f714e0:any[] = []
        _argv_e22ae90d_2428_3197_a8fb_549203f714e0.push(entity_id);
        let _cb_uuid = this.entity.call_hub_request("get_self_rank", encode(_argv_e22ae90d_2428_3197_a8fb_549203f714e0));
        return new get_rank_get_self_rank_cb(_cb_uuid, this.entity);
    }

    public  get_rank(start:number, end:number) {
        let _argv_e869f1c8_1f14_384f_aba6_2af2b54335e7:any[] = []
        _argv_e869f1c8_1f14_384f_aba6_2af2b54335e7.push(start);
        _argv_e869f1c8_1f14_384f_aba6_2af2b54335e7.push(end);
        let _cb_uuid = this.entity.call_hub_request("get_rank", encode(_argv_e869f1c8_1f14_384f_aba6_2af2b54335e7));
        return new get_rank_get_rank_cb(_cb_uuid, this.entity);
    }

}

File: sample/client/py/engine/get_rank_cli.py
from threading import Timer
from collections.abc import Callable
from enum import Enum
from .engine import *
from .engine.msgpack import *
from .common_cli import *

# this enum code is codegen by geese codegen for python

#this struct code is codegen by geese codegen for python
#this caller code is codegen by geese codegen for python
class get_rank_get_self_rank_cb(object):
    def __init__(self, _cb_uuid:int, _entity:subentity|player):
        self.entity = _entity
        self.cb:Callable[[role_rank_info], None] = None
        self.err:Callable[[error_code], None] = None
        self.rsp = callback(lambda: self.entity.del_callback(_cb_uuid))
        self.entity.reg_hub_callback(_cb_uuid, self.rsp)

    def on_rsp(self, bin:bytes):
        inArray = loads(bin)
        _rank = protcol_to_role_rank_info(inArray[0])
        self.cb(_rank)

    def on_err(self, bin:bytes):
        inArray = loads(bin)
        _err = inArray[0]
        self.err(_err)

    def callBack(self, _cb:Callable[[role_rank_info], None], _err:Callable[[error_code], None]):
        self.cb = _cb
        self.err = _err
        self.rsp.callback(self.on_rsp, self.on_err)
        return self.rsp

class get_rank_get_rank_cb(object):
    def __init__(self, _cb_uuid:int, _entity:subentity|player):
        self.entity = _entity
        self.cb:Callable[[list[role_rank_info]], None] = None
        self.err:Callable[[error_code], None] = None
        self.rsp = callback(lambda: self.entity.del_callback(_cb_uuid))
        self.entity.reg_hub_callback(_cb_uuid, self.rsp)

    def on_rsp(self, bin:bytes):
        inArray = loads(bin)
        _rank_list = []
        for v_e249ecbd_d64c_526b_901b_6d4ddee4b75a in inArray[0]:
            _rank_list.append(role_rank_info_to_protcol(v_e249ecbd_d64c_526b_901b_6d4ddee4b75a))
        self.cb(_rank_list)

    def on_err(self, bin:bytes):
        inArray = loads(bin)
        _err = inArray[0]
        self.err(_err)

    def callBack(self, _cb:Callable[[list[role_rank_info]], None], _err:Callable[[error_code], None]):
        self.cb = _cb
        self.err = _err
        self.rsp.callback(self.on_rsp, self.on_err)
        return self.rsp

class get_rank_caller(object):
    def __init__(self, entity:subentity|player):
        self.entity = entity

    def get_self_rank(self, entity_id:str):
        _argv_e22ae90d_2428_3197_a8fb_549203f714e0 = []
        _argv_e22ae90d_2428_3197_a8fb_549203f714e0.append(entity_id)
        _cb_uuid = self.entity.call_hub_request("get_self_rank", dumps(_argv_e22ae90d_2428_3197_a8fb_549203f714e0))

        return get_rank_get_self_rank_cb(_cb_uuid, self.entity)

    def get_rank(self, start:int, end:int):
        _argv_e869f1c8_1f14_384f_aba6_2af2b54335e7 = []
        _argv_e869f1c8_1f14_384f_aba6_2af2b54335e7.append(start)
        _argv_e869f1c8_1f14_384f_aba6_2af2b54335e7.append(end)
        _cb_uuid = self.entity.call_hub_request("get_rank", dumps(_argv_e869f1c8_1f14_384f_aba6_2af2b54335e7))

        return get_rank_get_rank_cb(_cb_uuid, self.entity)



